File: Creational/Builder/Python/Structure/builder.py
from abc import ABC, abstractmethod


class Builder(ABC):
    """
    The Builder interface specifies methods for creating the different parts
    of the Product objects.
    """
    
    @abstractmethod
    def build_part_a(self):
        """Build part A of the product"""
        pass
    
    @abstractmethod
    def build_part_b(self):
        """Build part B of the product"""
        pass
    
    @abstractmethod
    def build_part_c(self):
        """Build part C of the product"""
        pass
    
    @abstractmethod
    def get_result(self):
        """Return the built product"""
        pass

File: Creational/Builder/TypeScript/Structure/Builder.ts
import { Product } from "./Product";

/**
 * Builder - Abstract interface for creating products
 * Defines the contract for building complex objects step by step
 */
export abstract class Builder {
  protected product: Product = new Product();

  /**
   * Resets the builder to start fresh
   */
  reset(): void {
    this.product = new Product();
  }

  /**
   * Returns the built product
   */
  getProduct(): Product {
    const result = this.product;
    this.reset(); // Reset for next build
    return result;
  }

  /**
   * Abstract methods that concrete builders must implement
   */
  abstract buildPartA(): void;
  abstract buildPartB(): void;
  abstract buildPartC(): void;
}
